Routing Exon and Transcript construction through memoized methods

iskandr opened the pull request:
Make sure we never construct the same `Transcript` or `Exon` object twice by routing through `EnsemblRelease.exon_by_id` and `EnsemblRelease.transcript_by_id`. This not only cuts down on the cost of initialization but makes sure we don't recompute cached properties.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/pyensembl/84)
<!-- Reviewable:end -->

tavinathanson commented:
---

<sup>**[pyensembl/common.py, line 21 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFg1OhK6O8pdsJJxyO)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L21)):</sup>
Have you thought about turning this into its own repo, ala memoized-property?

---

<sup>**[pyensembl/common.py, line 30 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFfgmvTWENbugy2gOW)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L30)):</sup>
You removed the sort because you want a different order to be a different key?

---

<sup>**[pyensembl/common.py, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFfrteEft0AEkaTd_7)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L41)):</sup>
Is it really better not to blow up? Doesn't this mean that a programmer who thinks he's caching, isn't actually caching? Not clear why that's good?

---

<sup>**[pyensembl/common.py, line 47 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFjRCP0hiG8J8jAFEJ)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L47)):</sup>
Comment that says why this is necessary ("Needed to ensure that EnsemblRelease.clear_cache clears all these cached values")?

---

<sup>**[pyensembl/ensembl_release.py, line 133 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFiLzHKxb4E59ETsOy)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/ensembl_release.py#L133)):</sup>
I'm trying to think of alternatives to literally iterating through every possibility to check for a `clear_cache`, which feels somewhat hacky to me (in terms of the iteration, and also in terms of the reliance on `clear_cache` being named correctly; if it was accidentally named `clear_caches`, would we even see an issue?)

What about some form of "registering" a `clear_cache` function as a cache clearing function? With a `@clear_cache` decorator or something? And then iterating over everything registered? Just thinking out loud.

(tl;dr: registering functions that need clearing, with their associated clearing function, rather than using the existence of `clear_cache`?)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->

tavinathanson commented:
Reviewed files:
- pyensembl/common.py @ r1
- pyensembl/ensembl_release.py @ r1
- pyensembl/gene.py @ r1
- pyensembl/transcript.py @ r1

---



---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->

iskandr commented:
<sup>**[pyensembl/common.py, line 21 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFg1OhK6O8pdsJJxyO)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L21)):</sup>
Yeah, many times :-)  I want to do it once the two near-copies of @memoize between Varcode and PyEnsembl stabilize in their API/implementation.

---

<sup>**[pyensembl/common.py, line 30 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFfgmvTWENbugy2gOW)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L30)):</sup>
On second thought, this doesn't look right and I think the original cache_key was correct. 

---

<sup>**[pyensembl/common.py, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFfrteEft0AEkaTd_7)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L41)):</sup>
I wasn't sure what to do with unhashable but potentially common arg types like lists and dicts. I've switched this to check for hashability and using a __repr__ based key when an unhashable arg is encountered. What do you think?

---

<sup>**[pyensembl/ensembl_release.py, line 133 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFiLzHKxb4E59ETsOy)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/ensembl_release.py#L133)):</sup>
I'm not sure what that registering could look like, can you give a clearer example by any chance? (sorry for my lack of imagination)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->

tavinathanson commented:
Reviewed files:
- pyensembl/common.py @ r2
- pyensembl/database.py @ r2
- pyensembl/ensembl_release.py @ r3
- pyensembl/exon.py @ r3
- pyensembl/gene.py @ r2
- pyensembl/locus.py @ r3
- pyensembl/transcript.py @ r3
- test/test_exon_object.py @ r3
- test/test_gene_objects.py @ r2
- test/test_locus.py @ r2
- test/test_transcript_objects.py @ r2

---

<sup>**[pyensembl/common.py, line 26 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmssE_OAGnc7CLxJZtS)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L26)):</sup>
"any element of the cache key"

---

<sup>**[pyensembl/common.py, line 32 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmssq_rmAZevEHymPPf)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L32)):</sup>
It worries me to assume that `repr` + `type` are enough for a key, for anything that fails `hash`. That seems like more trust in `repr`'s output than is generally warranted?

What about having some specific logic for `lists` and such, and requiring any other object to be properly `hash`able.

---

<sup>**[pyensembl/common.py, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFfrteEft0AEkaTd_7)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/common.py#L41)):</sup>
Gotcha. Per my other comment, I'm thinking specific logic for lists and dicts over a blanket `repr` approach.

---

<sup>**[pyensembl/common.py, line 65 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmszy0YzcjBVpzuPct9)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L65)):</sup>
I'm confused about where you use this?

---

<sup>**[pyensembl/ensembl_release.py, line 133 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFiLzHKxb4E59ETsOy)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/ensembl_release.py#L133)):</sup>
I was mostly just thinking out loud, and no need to get too far into this, but e.g. within `gtf.py`:

```
@register('clear')
def clear_cache(self):
   self._dataframes.clear()
   clear_cached_objects()
```

And in `registry.py`:

```
def register(registry_name, func):
   @wraps(fn)
   def wrapped_fn(...):
     registry[registry_name].add(fn)
  ...
```

And then, instead of:

```
for maybe_fn in self.__dict__.values():
  if hasattr(maybe_fn, 'clear_cache'):
    maybe_fn.clear_cache()
````

Just:

```
for fn in registry['clear']:
  fn()
```

...or something.

---

<sup>**[pyensembl/exon.py, line 25 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt1mikQP_1uV1ny5yO)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/pyensembl/exon.py#L25)):</sup>
`require_string` still seems relevant for this one, `gene.py`, etc.?

---

<sup>**[pyensembl/exon.py, line 114 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt1wqzp0JtJQHFD0rX)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/pyensembl/exon.py#L114)):</sup>
I''m confused about where this went?

---

<sup>**[pyensembl/transcript.py, line 52 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt04ZTgB-VHNM7kYy2)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/transcript.py#L52)):</sup>
What happened to all the checking? Edit: ah, the point is that happens upstream.

---

<sup>**[pyensembl/transcript.py, line 75 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-yBAqV0lTDxpuaIH)** ([raw file](https://github.com/hammerlab/pyensembl/blob/c92b340af080575e636571eac96b141c5fb0aa5d/pyensembl/transcript.py#L75)):</sup>
I assume this was also supposed to be uncommented?

---

<sup>**[pyensembl/transcript.py, line 269 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-VgDxn_mrfwy1CbC)** ([raw file](https://github.com/hammerlab/pyensembl/blob/c92b340af080575e636571eac96b141c5fb0aa5d/pyensembl/transcript.py#L269)):</sup>
What's the deal with this commented line? I assume it was intended to be uncommented?

---

<sup>**[test/test_exon_object.py, line 76 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-LVK8YeDJhQ0gQlV)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/test/test_exon_object.py#L76)):</sup>
What happened to these tests?

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->

iskandr commented:
<sup>**[pyensembl/common.py, line 32 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmssq_rmAZevEHymPPf)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L32)):</sup>
`repr` is *supposed* to give you an invertible representation but of course that won't always be true. I guess this sort of concern is why I'm hesitant to make `memoize` a standalone library. 

I'm not totally thrilled about making `list` and `dict` a special case since it's not hard to imagine wanting a `dict` which has `list` values or vice versa. I extended the docstring to talk about the fragile assumption I'm making about type(arg) and repr(arg) identifying the argument. 

---

<sup>**[pyensembl/common.py, line 65 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmszy0YzcjBVpzuPct9)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L65)):</sup>
When one method (`EnsemblRelease.exons_of_transcript_id`) may need to update the cache of another (`EnsemblRelease.exon_by_id`). It'll be messy but not sure how to make it cleaner. 

---

<sup>**[pyensembl/ensembl_release.py, line 133 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFiLzHKxb4E59ETsOy)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/ensembl_release.py#L133)):</sup>
Haven't we now just replaced the problem with needing to remember to register these functions and make sure they go into the correct registry? Seems just as error prone to me. 

(also global function registries feel icky)

---

<sup>**[pyensembl/exon.py, line 25 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt1mikQP_1uV1ny5yO)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/pyensembl/exon.py#L25)):</sup>
I'm accepting that Python isn't yet a statically typed language. A lot of these `require_string` usages, particularly those for constructors normally called from within PyEnsembl, are probably excessive. 

---

<sup>**[pyensembl/exon.py, line 114 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt1wqzp0JtJQHFD0rX)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/pyensembl/exon.py#L114)):</sup>
Deleted. It was nonsensical to talk about the start & stop codons of an exon. They don't have codons at all, or a reading frame. It's up to `Transcript` to handle this kind of information.

---

<sup>**[pyensembl/transcript.py, line 75 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-yBAqV0lTDxpuaIH)** ([raw file](https://github.com/hammerlab/pyensembl/blob/c92b340af080575e636571eac96b141c5fb0aa5d/pyensembl/transcript.py#L75)):</sup>
Left behind by mistake

---

<sup>**[pyensembl/transcript.py, line 269 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-VgDxn_mrfwy1CbC)** ([raw file](https://github.com/hammerlab/pyensembl/blob/c92b340af080575e636571eac96b141c5fb0aa5d/pyensembl/transcript.py#L269)):</sup>
Sorry, the usage of exon_intervals was an experiment that I forgot to clean up. I found using the Exon objects directly to be clearer. 

---

<sup>**[test/test_exon_object.py, line 76 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt-LVK8YeDJhQ0gQlV)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/test/test_exon_object.py#L76)):</sup>
See above (they were testing properties that I think Exons shouldn't have)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->

tavinathanson commented:
Looks good!

---

Reviewed files:
- pyensembl/common.py @ r4
- pyensembl/transcript.py @ r4

---

<sup>**[pyensembl/common.py, line 32 \[r2\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmssq_rmAZevEHymPPf)** ([raw file](https://github.com/hammerlab/pyensembl/blob/bfef452aed6e48ffd1a56b30f92156c6595e4fe8/pyensembl/common.py#L32)):</sup>
Fair enough!

---

<sup>**[pyensembl/ensembl_release.py, line 133 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-JmFiLzHKxb4E59ETsOy)** ([raw file](https://github.com/hammerlab/pyensembl/blob/a6402f3c4a793adb82e5bc82ab9518f5465aca96/pyensembl/ensembl_release.py#L133)):</sup>
Well, it feels a bit more explicit to me. Relying on naming the function `clear_cache` vs. remembering to decorate it are slightly different. That being said, I'm by no means prepared to defend this approach ;)

---

<sup>**[pyensembl/exon.py, line 25 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pyensembl/84#-Jmt1mikQP_1uV1ny5yO)** ([raw file](https://github.com/hammerlab/pyensembl/blob/e13344499d6e7e751d28e89d52021454b3ed39f5/pyensembl/exon.py#L25)):</sup>
It's fascinating watching your approach to types evolve :)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pyensembl/84)
<!-- Sent from Reviewable.io -->
